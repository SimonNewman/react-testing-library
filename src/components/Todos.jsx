"use client";

import { useState } from "react";
import Button from "./Button";

const Todos = () => {
  const [todoValue, setTodoValue] = useState("");
  const [todos, setTodos] = useState([]);

  const handleChange = (e) => {
    setTodoValue(e.target.value);
  };

  const handleAddTodo = (e) => {
    e.preventDefault();

    if (todoValue.trim()) {
      setTodos((prevTodos) => [...prevTodos, todoValue.trim()]);
      setTodoValue("");
    }
  };

  return (
    <>
      <label>
        New Todo
        <input
          className="border border-black text-black"
          onChange={handleChange}
          value={todoValue}
        />
      </label>

      <Button onClick={handleAddTodo}>Add Todo</Button>

      {todos.length > 0 && (
        <ul>
          {todos.map((todo, index) => (
            <li key={`${todo}-${index}`}>{todo}</li>
          ))}
        </ul>
      )}
    </>
  );
};

export default Todos;
