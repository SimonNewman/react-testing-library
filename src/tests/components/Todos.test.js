import "@testing-library/jest-dom";

import { fireEvent, render, screen } from "@testing-library/react";

import Todos from "@/components/Todos";
import renderer from "react-test-renderer";

const newTodoText = "My first todo";

it("Renders Button component and simulates adding a todo", async () => {
  // Render the Todos component
  render(<Todos />);

  // Get the input and button elements
  const input = screen.getByLabelText("New Todo");
  const button = screen.getByRole("button");

  // Simulate user typing new todo
  fireEvent.change(input, { target: { value: newTodoText } });

  // Simulate a user clicking the button
  fireEvent.click(button);

  // Check if the new todo was added to the list
  const todos = screen.getByRole("list");
  expect(todos).toHaveTextContent(newTodoText);

  // Check that todo input has been cleared
  expect(input).toHaveTextContent("");

  // Snapshot
  const tree = renderer.create(<Todos />).toJSON();
  expect(tree).toMatchSnapshot();
});
