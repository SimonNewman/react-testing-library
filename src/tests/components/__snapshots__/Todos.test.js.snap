// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Button component and simulates adding a todo 1`] = `
[
  <label>
    New Todo
    <input
      className="border border-black text-black"
      onChange={[Function]}
      value=""
    />
  </label>,
  <button
    onClick={[Function]}
  >
    Add Todo
  </button>,
]
`;
